### PR TITLE
[serve] Implement `target_capacity`

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Implements the `target_capacity` functionality to adjust the number of replicas across all deployments in the cluster.

The desired semantics for autoscaling are to scale the bounds (`min_replicas` and `max_replicas`) of the autoscaling algorithm as well as `initial_num_replicas`. This is implemented by running the unmodified autoscaling algorithm to set the `target_info.num_replicas`, then always scaling that following a singular codepath.

Before merging:

- [x] Unit tests
- [ ] Integration tests

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/tests/unit/test_deployment_state.py [hunk 1]
```diff
@@ -2701,5 +2701,440 @@ def test_get_active_node_ids_none(mock_deployment_state_manager_full):
...
+            {}, {deployment_state._id: deployment_state_update_result.downscale}
+        )[deployment_state._id]
+        deployment_state.stop_replicas(replicas_to_stop)
+        check_counts(
+            deployment_state,
+            total=10,
+            by_state=[(ReplicaState.RUNNING, 5), (ReplicaState.STOPPING, 5)],
+        )
+
+        # TODO(edoakes): when we update the state machine to include
```
Comment (edoakes) [inline python/ray/serve/tests/unit/test_deployment_state.py:2870]: @zcin

File: python/ray/serve/_private/deployment_state.py [hunk 2]
```diff
@@ -1398,6 +1398,21 @@ def _set_target_state_autoscaling(self, num_replicas: int) -> None:
         self._save_checkpoint_func(writeahead_checkpoints={self._id: target_state})
         self._target_state = target_state
 
+    @staticmethod
+    def get_capacity_adjusted_num_replicas(
+        num_replicas: int, target_capacity: Optional[float]
+    ) -> int:
+        """Return the target state `num_replicas` adjusted by the `target_capacity`.
+
+        The output will never be lower than `1`.
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/deployment_state.py:1408]: > The output will never be lower than `1`.

What does this mean for scale-to-zero? Can deployments be scaled to zero if the config has a `target_capacity`?

File: python/ray/serve/_private/deployment_state.py [hunk 3]
```diff
@@ -1478,10 +1498,20 @@ def autoscale(self, current_handle_queued_queries: int):
...
+            self._target_state.num_replicas, target_capacity
+        )
+        adjusted_decision_num_replicas = self.get_capacity_adjusted_num_replicas(
+            decision_num_replicas, target_capacity
+        )
         logger.info(
             f"Autoscaling replicas for deployment {self.deployment_name} in "
-            f"application {self.app_name} from {self._target_state.num_replicas} to "
-            f"{decision_num_replicas}. Current ongoing requests: "
+            f"application {self.app_name} from {adjusted_curr_num_replicas} to "
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/deployment_state.py:1519]: > from {adjusted_curr_num_replicas}

If `target_capacity` was changed just before this iteration, then the `adjusted_curr_num_replicas` may not match the actual number of currently running replicas. I think we should use the `num_replicas` directly in this log message.

File: python/ray/serve/_private/deployment_state.py [hunk 4]
```diff
@@ -2449,11 +2487,20 @@ def get_handle_queueing_metrics(
...
         return current_handle_queued_queries
 
-    def update(self) -> bool:
+    def update(self, target_capacity: Optional[float] = None) -> bool:
         """Updates the state of all deployments to match their goal state.
 
+        `target_capacity` represents the target capacity percentage for all replicas
+        across the cluster. The `num_replicas`, `min_replicas`, and `max_replicas` for
+        each deployment will be scaled by this percentage.
+
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/deployment_state.py:2502]:
```suggestion
        `target_capacity` represents the target capacity percentage for all deployments
        across the cluster. The `num_replicas`, `min_replicas`, and `max_replicas` for
        each deployment will be scaled by this percentage.

```

File: python/ray/serve/_private/deployment_state.py [hunk 5]
```diff
@@ -2449,11 +2487,20 @@ def get_handle_queueing_metrics(
...
         Returns True if any of the deployments have replicas in the RECOVERING state.
         """
+        if target_capacity is not None and (
+            target_capacity < 0 or target_capacity > 100
+        ):
+            raise ValueError("`target_capacity` must be between 0 and 100.")
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/deployment_state.py:2508]:
```suggestion
            raise ValueError(f"Got invalid `target_capacity` {target_capacity}. `target_capacity` must be between 0 and 100.")
```

File: python/ray/serve/tests/unit/test_deployment_state.py [hunk 6]
```diff
@@ -2701,5 +2701,440 @@ def test_get_active_node_ids_none(mock_deployment_state_manager_full):
     assert None not in deployment_state_manager.get_active_node_ids()
 
 
+class TestTargetCapacity:
+    """
+    Tests related to the `target_capacity` field that adjusts the target num_replicas.
+    """
+
+    @pytest.mark.parametrize(
+        "num_replicas,target_capacity,expected_output",
```
Comment (shrekris-anyscale) [inline python/ray/serve/tests/unit/test_deployment_state.py:2710]: Could you add a test case where the `num_replicas` is 0?

File: python/ray/serve/tests/unit/test_deployment_state.py [hunk 7]
```diff
@@ -2701,5 +2701,440 @@ def test_get_active_node_ids_none(mock_deployment_state_manager_full):
...
+        assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+
+        # Now update target_capacity back to None, should have no effect.
+        deployment_state.update(target_capacity=None)
+        check_counts(deployment_state, total=2, by_state=[(ReplicaState.RUNNING, 2)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+
+    def test_target_capacity_0_1_replica(self, mock_deployment_state):
+        """
+        Deploy with target_capacity set to 0. Should have a single replica.
```
Comment (shrekris-anyscale) [inline python/ray/serve/tests/unit/test_deployment_state.py:2808]: If `target_capacity` is 0, I'd expect there to be 0 replicas because we'd be scaling `num_replicas` and the `autoscaling_bounds` down to 0. Is there a reason 0 is treated as a special case?

File: python/ray/serve/tests/unit/test_deployment_state.py [hunk 8]
```diff
@@ -2701,5 +2701,440 @@ def test_get_active_node_ids_none(mock_deployment_state_manager_full):
...
+
+        deployment_state.update(target_capacity=0)
+        check_counts(deployment_state, total=1, by_state=[(ReplicaState.RUNNING, 1)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+
+        deployment_state.update(target_capacity=0)
+        check_counts(deployment_state, total=1, by_state=[(ReplicaState.RUNNING, 1)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+
+    def test_reduce_target_capacity(self, mock_deployment_state):
```
Comment (shrekris-anyscale) [inline python/ray/serve/tests/unit/test_deployment_state.py:2834]: Could you add a test case that switches directions? E.g. starts by reducing and then increases?

**Review by shrekris-anyscale — COMMENTED** [review]
Nice work!

File: python/ray/serve/tests/unit/test_deployment_state.py (hunk 8 above)
Comment (zcin) [inline python/ray/serve/tests/unit/test_deployment_state.py:2834]: What is an example of what target capacity would look like if a rollback happens? Can add a test for that?

**Review by zcin — COMMENTED** [review]
Cool!

File: python/ray/serve/_private/deployment_state.py [hunk 9]
```diff
@@ -1398,6 +1399,26 @@ def _set_target_state_autoscaling(self, num_replicas: int) -> None:
         self._save_checkpoint_func(writeahead_checkpoints={self._id: target_state})
         self._target_state = target_state
 
+    @staticmethod
+    def get_capacity_adjusted_num_replicas(
+        num_replicas: int, target_capacity: Optional[float]
+    ) -> int:
+        """Return the target state `num_replicas` adjusted by the `target_capacity`.
+
+        The output will never be lower than `1`.
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/deployment_state.py:1408]: fixed

File: python/ray/serve/_private/deployment_state.py (hunk 2 above)
Comment (edoakes) [reply python/ray/serve/_private/deployment_state.py:1408]: good catch. fixed and added test case for scale-to-zero.

File: python/ray/serve/tests/unit/test_deployment_state.py (hunk 7 above)
Comment (edoakes) [reply python/ray/serve/tests/unit/test_deployment_state.py:2808]: This is to ensure our invariant that all requests can always succeed. I've fixed it for scale-to-zero, but for deployments with a fixed num_replicas, they should still always have at least one replica.

Comment (edoakes) [reply python/ray/serve/tests/unit/test_deployment_state.py:2808]: Otherwise `target_capacity=0` is equivalent to just deleting all applications

File: python/ray/serve/_private/deployment_state.py [hunk 10]
```diff
@@ -2034,14 +2065,23 @@ def _stop_replicas_on_draining_nodes(self):
             else:
                 self._replicas.add(replica.actor_details.state, replica)
 
-    def update(self) -> DeploymentStateUpdateResult:
+    def update(
+        self, target_capacity: Optional[float] = None
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/deployment_state.py:2069]: I much prefer moving in the direction of these functions being pure. I found it quite hard to follow the existing logic between autoscaling and the rest of the reconciliation because `.autoscale()` was overwriting the `target_info`, which I assumed was immutable and represented the target config.

In general testing and reasoning about the code is easier for pure functions.

[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/deployment_state.py:2069]: `self._target_state.num_replicas` is now only being used/adjusted at (1) the top level when calculating the adjusted version and (2) in `autoscale()` (which I think we should fix)

File: python/ray/serve/_private/deployment_state.py [hunk 11]
```diff
@@ -2055,9 +2095,13 @@ def update(self) -> DeploymentStateUpdateResult:
...
 
-            upscale, downscale = self._scale_deployment_replicas()
+            upscale, downscale = self._scale_deployment_replicas(
+                adjusted_target_num_replicas
+            )
 
-            deleted, any_replicas_recovering = self._check_curr_status()
+            deleted, any_replicas_recovering = self._check_curr_status(
+                adjusted_target_num_replicas
+            )
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/deployment_state.py:2104]: good catch fixed

File: python/ray/serve/_private/deployment_state.py [hunk 12]
```diff
@@ -2449,11 +2493,20 @@ def get_handle_queueing_metrics(
             current_handle_queued_queries = 0
         return current_handle_queued_queries
 
-    def update(self) -> bool:
+    def update(self, target_capacity: Optional[float] = None) -> bool:
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/deployment_state.py:2496]: In the future when we do in-cluster rollouts, `None` will mean that there's no ongoing rollout

File: python/ray/serve/tests/unit/test_deployment_state.py [hunk 13]
```diff
@@ -2701,5 +2701,440 @@ def test_get_active_node_ids_none(mock_deployment_state_manager_full):
...
+        deployment_state.update(target_capacity=100)
+        check_counts(deployment_state, total=10, by_state=[(ReplicaState.RUNNING, 10)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+
+        # Reduce target_capacity to 50, half the replicas should be stopped.
+        deployment_state_update_result = deployment_state.update(target_capacity=50)
+        replicas_to_stop = deployment_state._deployment_scheduler.schedule(
+            {}, {deployment_state._id: deployment_state_update_result.downscale}
+        )[deployment_state._id]
+        deployment_state.stop_replicas(replicas_to_stop)
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/tests/unit/test_deployment_state.py:2863]: Oh that makes sense, I was looking for something like this. The way these tests are written is now very error prone :( -- if you missed the `schedule` call, you might accidentally not be testing the full e2e behavior.

File: python/ray/serve/tests/unit/test_deployment_state.py (hunk 8 above)
Comment (edoakes) [reply python/ray/serve/tests/unit/test_deployment_state.py:2834]: For now, we are not supporting in-place rollouts within the cluster, so it would just be reducing target_capacity for the target cluster and increasing it for the existing cluster (same path as the opposite direction)

Comment (edoakes) [reply python/ray/serve/tests/unit/test_deployment_state.py:2834]: `test_target_capacity_with_changing_num_replicas` does this, but really there's no need for it; the reconciler is stateless and has no notion of directionality at the moment.

File: python/ray/serve/_private/deployment_state.py (hunk 3 above)
Comment (edoakes) [reply python/ray/serve/_private/deployment_state.py:1519]: This is a good point. Perhaps it's best to just leave off the "from" portion of the log message? I think it'll be pretty confusing if the autoscaler says we're scaling to 8 replicas but then we are happy at 4 due to a target_capacity of 50.

Comment (shrekris-anyscale) [reply python/ray/serve/_private/deployment_state.py:1524]: Yeah, I'm okay with that.

File: python/ray/serve/tests/unit/test_deployment_state.py (hunk 7 above)
Comment (shrekris-anyscale) [reply python/ray/serve/tests/unit/test_deployment_state.py:2812]: > Otherwise `target_capacity=0` is equivalent to just deleting all applications

I think that make sense. `target_capacity=0` is equivalent to scaling the target version to 0. Imo, the invariant should be that `target_capacity>0` means at least 1 replica for each (non-scale-to-zero) deployment should be running. And this implementation already supports that because it rounds up when doing `num_replicas` calculations.

Comment (edoakes) [reply python/ray/serve/tests/unit/test_deployment_state.py:2812]: There's no point to ever using `target_capacity=0` then, because you might as well just delete everything (if scaling down) or make your first step `>0` (if scaling up).

Also, if this cannot be used to express the minimal scale of the applications, there's no other semantically clear way to do it. Do you pick `10`? Then if you have >10 replicas, it doesn't work. Same for `1` (`>100` replicas) or `0.1` (`>1000` replicas).

Makes things messy.

Comment (shrekris-anyscale) [reply python/ray/serve/tests/unit/test_deployment_state.py:2812]: > Also, if this cannot be used to express the minimal scale of the applications, there's no other semantically clear way to do it.

That's a good point I hadn't considered. I'm okay with keeping the invariant as-is: there's always at least 1 replica for every deployment.

**Review by shrekris-anyscale — APPROVED** [review]
Nice job, this change looks good!

File: python/ray/serve/tests/test_target_capacity.py [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+        check_expected_num_replicas,
+        deployment_to_num_replicas={
+            SCALE_TO_ZERO_DEPLOYMENT_NAME: SCALE_TO_ZERO_DEPLOYMENT_MAX_REPLICAS / 2,
+        },
+    )
+
+    # TODO(edoakes): for some reason, the deployment does not actually scale down to
+    # zero here, so skipping this part for now. Instead it repeatedly scales down and
+    # then back up. Seems to have something to do with the handle-side queue metric.
+    return
```
Comment (edoakes) [inline python/ray/serve/tests/test_target_capacity.py:300]: @zcin could you have a look at this when you have a chance? I tried debugging for awhile but couldn't figure it out. It seemed that even when I increased the metrics collection interval on the handle, the handle queue metric was just stuck a `1` throughout the entire test.

Going to file a follow-up issue for it so it isn't lost.
